src/stdlib_sorting_ord_sort.fypp

```diff
@@ -0,0 +1,738 @@
+            ${t1}$, intent(inout) :: array(0:)
+
+            integer(int_size) :: lo, hi
+            ${t1}$ :: temp
+
+            lo = 0
+            hi = size( array, kind=int_size ) - 1
+            do while( lo < hi )
+                temp = array(lo)
+                array(lo) = array(hi)
+                array(hi) = temp
+                lo = lo + 1
+                hi = hi - 1
+            end do
+
+        end subroutine reverse_segment
+
+    end subroutine ${k1}$_ord_sort
+
+#:endfor
+
+
+    module subroutine char_ord_sort( array, work )
+! A translation to Fortran 2008, of the `"Rust" sort` algorithm found in
+! `slice.rs`
+! https://github.com/rust-lang/rust/blob/90eb44a5897c39e3dff9c7e48e3973671dcd9496/src/liballoc/slice.rs#L2159
+! The Rust version in turn is a simplification of the Timsort algorithm
+! described in
+! https://svn.python.org/projects/python/trunk/Objects/listsort.txt, as
+! it drops both the use of 'galloping' to identify bounds of regions to be
+! sorted and the estimation of the optimal `run size`. However it remains
+! a hybrid sorting algorithm combining an iterative Merge sort controlled
+! by a stack of `RUNS` identified by regions of uniformly decreasing or
+! non-decreasing sequences that may be expanded to a minimum run size and
+! initially processed by an insertion sort.
+!
+! Note the Fortran implementation simplifies the logic as it only has to
+! deal with Fortran arrays of intrinsic types and not the full generality
+! of Rust's arrays and lists for arbitrary types. It also adds the
+! estimation of the optimal `run size` as suggested in Tim Peters'
+! original `listsort.txt`, and an optional `work` array to be used as
+! scratch memory.
+        character(len=*), intent(inout)                    :: array(0:)
+        character(len=len(array)), intent(inout), optional :: work(0:)
+
+        character(len=:), allocatable :: buf(:)
+        integer(int_size) :: array_size
+        integer :: stat
+
+        if ( present(work) ) then
+! Use the work array as scratch memory
+            call merge_sort( array, work )
+        else
+! Allocate a buffer to use as scratch memory.
+            array_size = size( array, kind=int_size )
+            allocate( character(len=len(array)) :: buf(0:array_size/2-1), &
+                      stat=stat )
+            if ( stat /= 0 ) error stop "${k1}$_ord_sort: Allocation of buffer failed."
+            call merge_sort( array, buf )
+        end if
+
+    contains
+
+        pure function calc_min_run( n ) result(min_run)
+!! Returns the minimum length of a run from 32-63 so that N/MIN_RUN is
+!! less than or equal to a power of two. See
+!! https://svn.python.org/projects/python/trunk/Objects/listsort.txt
+            integer(int_size)             :: min_run
+            integer(int_size), intent(in) :: n
+
+            integer(int_size) :: num, r
+
+            num = n
+            r = 0_int_size
+
+            do while( num >= 64 )
```

Comment:
Is the value `64` depending on the kind of `r` (i.e., `int_size`)? What happen is `int_size` become `int32`?

Reply:
No the 64 is independent of the kind of integer. It is an upper limit on the block size to be used for insertion sorts prior to using the heap sort. It depends on the relative processing speed of insertion sort versus heap sort. Ideally you divide the input array into a power of two number of blocks, but it is better to have sightly less than a power of two than slightly more. Tim Peters did extensive testing with `num >= 16`, `num >= 32`, `num >= 64`, and `num >= 128`, and found that the performance was fairly flat for the different cutoffs. Of course his testing was with Python, and the ideal cutoff may be language dependent. Once we have the API fixed we could run our own tests for different cutoffs and see which cutoff is best.

Reply:
Thank you for the explanation.